### PR TITLE
[IMPROVEMENT] User Can Copy Link On hold

### LINKED ISSUE #1680: [IMPROVEMENT] Hold to copy link
Users should be able to hold links and copy to clipboard.

### PR DESCRIPTION
**Pull request by sarthakpranesh**
@RocketChat/ReactNative
Closes #1680

URLs sent as the message can now be copied on long pressing or holding the message.

### CODE REVIEW

File: app/containers/message/Urls.js [hunk 1]
```diff
@@ -81,10 +88,19 @@ const Url = React.memo(({
...
-
+	const onLongPress = () => {
+		Clipboard.setString(url.url);
+		if (Platform.os === 'ios') {
+			Vibration.vibrate();
+		} else {
+			Vibration.vibrate(200);
+			ToastAndroid.show('Link copied', ToastAndroid.SHORT);
+		}
+	};
```
Comment (djorkaeffalexandre) [inline app/containers/message/Urls.js]: For first, you don't need to use `ToastAndroid`, we have our own [Toast](https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/containers/Toast.js) Component, you can search on our code how to use this.
I don't think you need to add some vibration when a link was copied, only the Toast is a feedback great.

Comment (sarthakpranesh) [reply app/containers/message/Urls.js]: Ok, Thanks for the guidance. :+1:

File: app/containers/message/Urls.js [hunk 2]
```diff
@@ -1,10 +1,13 @@
 import React from 'react';
-import { View, Text, StyleSheet } from 'react-native';
+import {
+	View,
+	Text,
+	StyleSheet
+} from 'react-native';
```
Comment (djorkaeffalexandre) [inline app/containers/message/Urls.js]: Remove these changes.

File: app/containers/message/Urls.js [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
 import PropTypes from 'prop-types';
 import FastImage from 'react-native-fast-image';
 import Touchable from 'react-native-platform-touchable';
 import isEqual from 'lodash/isEqual';
-
```
Comment (djorkaeffalexandre) [inline app/containers/message/Urls.js]: Undo these changes.

File: app/containers/message/Urls.js [hunk 4]
```diff
@@ -81,7 +84,6 @@ const Url = React.memo(({
 	}
 
 	const onPress = () => openLink(url.url, theme);
-
```
Comment (djorkaeffalexandre) [inline app/containers/message/Urls.js]: Undo these changes.

File: app/containers/markdown/Link.js [hunk 5]
```diff
@@ -18,10 +21,16 @@ const Link = React.memo(({
...
+	const onLongPress = () => {
+		Clipboard.setString(link);
+		EventEmitter.emit(LISTENER, { message: I18n.t('Copied_to_clipboard') });
+	};
+
 	// if you have a [](https://rocket.chat) render https://rocket.chat
 	return (
 		<Text
 			onPress={preview ? undefined : handlePress}
+			onLongPress={onLongPress}
```
Comment (djorkaeffalexandre) [inline app/containers/markdown/Link.js]: You need to handle preview cases.

Comment (sarthakpranesh) [reply app/containers/markdown/Link.js]: Preview Cases are handled by the onPress function already.

File: app/containers/message/Urls.js (hunk 2 above)
Comment (sarthakpranesh) [reply app/containers/message/Urls.js]: I included Clipboard, Vibrations, Platform modules of react-native which are not required by Url.js as they aren't required there. I included them for fixing the issue but you redirected me to our own Toast.js component, therefore, I removed the modules.
I'll revert them.

**Comment by sarthakpranesh** [comment]
When adding LISTENER from Toast.js and EventEmiter from events and using them for Toasts in markdown/Link.js the following error is thrown.
image

File: app/containers/message/Urls.js [hunk 6]
```diff
@@ -1,15 +1,24 @@
...
+	StyleSheet,
+	Clipboard,
+	ToastAndroid,
+	Platform,
+	Vibration
+} from 'react-native';
 import PropTypes from 'prop-types';
 import FastImage from 'react-native-fast-image';
 import Touchable from 'react-native-platform-touchable';
 import isEqual from 'lodash/isEqual';
```
Comment (djorkaeffalexandre) [inline app/containers/message/Urls.js]: Please, undo all changes in this file.

File: app/containers/markdown/Link.js (hunk 5 above)
Comment (djorkaeffalexandre) [reply app/containers/markdown/Link.js]: No, we not want that when user long press a link on roomsListView this link are copy. Use some condition like on `onPress`.

Comment (sarthakpranesh) [reply app/containers/markdown/Link.js]: Ok, I will apply some conditions

File: app/containers/message/Urls.js (hunk 6 above)
Comment (sarthakpranesh) [reply app/containers/message/Urls.js]: Yeah, they'll be undone

**Comment by sarthakpranesh** [comment]
Updates
1. relevant mocks are added to __mocks__
2. all changes made to the Url.js were undone ( as asked requested earlier )
3. app/containers/markdown/Link.js updated, condition applied to make sure the user does not copy the link from roomsListView by mistake onLongPress

Please let me know if there are any further changes to be made.
